src/Runner.Listener/SelfUpdaterV2.cs

```diff
@@ -0,0 +1,569 @@
+
+        public async Task<bool> SelfUpdate(RunnerRefreshMessage updateMessage, IJobDispatcher jobDispatcher, bool restartInteractiveRunner, CancellationToken token)
+        {
+            Busy = true;
+            try
+            {
+                var totalUpdateTime = Stopwatch.StartNew();
+
+                Trace.Info($"An update is available.");
+                _updateTrace.Enqueue($"RunnerPlatform: {updateMessage.OS}");
+
+                // Print console line that warn user not shutdown runner.
+                _terminal.WriteLine("Runner update in progress, do not shutdown runner.");
+                _terminal.WriteLine($"Downloading {updateMessage.TargetVersion} runner");
+
+                await DownloadLatestRunner(token, updateMessage.TargetVersion, updateMessage.DownloadUrl, updateMessage.SHA256Checksum, updateMessage.OS);
+                Trace.Info($"Download latest runner and unzip into runner root.");
+
+                // wait till all running job finish
+                _terminal.WriteLine("Waiting for current job finish running.");
+
+                await jobDispatcher.WaitAsync(token);
+                Trace.Info($"All running job has exited.");
+
+                // We need to keep runner backup around for macOS until we fixed https://github.com/actions/runner/issues/743
+                // delete runner backup
+                var stopWatch = Stopwatch.StartNew();
+                DeletePreviousVersionRunnerBackup(token, updateMessage.TargetVersion);
+                Trace.Info($"Delete old version runner backup.");
+                stopWatch.Stop();
+                // generate update script from template
+                _updateTrace.Enqueue($"DeleteRunnerBackupTime: {stopWatch.ElapsedMilliseconds}ms");
+                _terminal.WriteLine("Generate and execute update script.");
+
+                string updateScript = GenerateUpdateScript(restartInteractiveRunner, updateMessage.TargetVersion);
+                Trace.Info($"Generate update script into: {updateScript}");
+
+
+#if DEBUG
+                // For L0, we will skip execute update script.
+                if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("_GITHUB_ACTION_EXECUTE_UPDATE_SCRIPT")))
+                {
+                    string flagFile = "update.finished";
+                    IOUtil.DeleteFile(flagFile);
+                    // kick off update script
+                    Process invokeScript = new();
+#if OS_WINDOWS
+                    invokeScript.StartInfo.FileName = WhichUtil.Which("cmd.exe", trace: Trace);
+                    invokeScript.StartInfo.Arguments = $"/c \"{updateScript}\"";
+#elif (OS_OSX || OS_LINUX)
+                    invokeScript.StartInfo.FileName = WhichUtil.Which("bash", trace: Trace);
+                    invokeScript.StartInfo.Arguments = $"\"{updateScript}\"";
+#endif
+                    invokeScript.Start();
+                    Trace.Info($"Update script start running");
+                }
+#endif
```

Comment:
```suggestion
#if DEBUG
                // For L0, we will skip execute update script.
                if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("_GITHUB_ACTION_EXECUTE_UPDATE_SCRIPT")))
#endif
                {
                    string flagFile = "update.finished";
                    IOUtil.DeleteFile(flagFile);
                    // kick off update script
                    Process invokeScript = new();
#if OS_WINDOWS
                    invokeScript.StartInfo.FileName = WhichUtil.Which("cmd.exe", trace: Trace);
                    invokeScript.StartInfo.Arguments = $"/c \"{updateScript}\"";
#elif (OS_OSX || OS_LINUX)
                    invokeScript.StartInfo.FileName = WhichUtil.Which("bash", trace: Trace);
                    invokeScript.StartInfo.Arguments = $"\"{updateScript}\"";
#endif
                    invokeScript.Start();
                    Trace.Info($"Update script start running");
                }
```

I think we only want to skip the script if we are under debug and `_GITHUB_ACTION_EXECUTE_UPDATE_SCRIPT` is set to a non-empty string.

Please double check in case my brain was not working properly. 🤞

Reply:
good call. I had it reversed in my brain 😄